Race condition in redis broker `join`

# Issues

GitHub issues are for bugs.  If you have questions, please ask them on the [discussion board](https://reddit.com/r/dramatiq).

## Checklist

* [x] Does your title concisely summarize the problem?
* [x] Did you include a minimal, reproducible example?
* [x] What OS are you using?
* [x] What version of Dramatiq are you using?
* [x] What did you do?
* [x] What did you expect would happen?
* [x] What happened?


## What OS are you using?

<!-- for example: Ubuntu 16.04 or macOS 10.13.3 -->
windows 10 / ubuntu 20.04

## What version of Dramatiq are you using?

<!-- run this command to find out: python -c 'import dramatiq; print(dramatiq.__version__)' -->
1.12.0 and 1.12.3

## What did you do?

<!-- be descriptive, but succint -->
A test in an automated pipeline tests that an actor enqueued with a delay is executed after the `join` of the broker returns.

## What did you expect would happen?

<!-- be descriptive, but succint -->
The actor is correctly run

## What happened?

<!-- be descriptive, but succint -->
Sometimes the actor does not run, because of a race condition.

Following it's an example code can reproduces the race condition:
```py
import os
import dramatiq
import pytest
from dramatiq.brokers.redis import RedisBroker
from dramatiq.middleware import AgeLimit, Retries, ShutdownNotifications
import pathlib
import logging

logging.basicConfig()
logging.getLogger("dramatiq").setLevel(logging.DEBUG)

_MIDDLEWARES = [AgeLimit(), ShutdownNotifications(), Retries(max_retries=3)]
redis_broker = RedisBroker(
    host="localhost",
    middleware=_MIDDLEWARES,
    namespace=f"ns-{os.getpid()}",
    password="redis-passwd",
)
dramatiq.set_broker(redis_broker)

@dramatiq.actor
def bar(s):
    pathlib.Path(f"file{s}.canary").touch(exist_ok=False)
    print(s)

@pytest.fixture(scope="module")
def wM():
    w = dramatiq.Worker(redis_broker)
    w.start()
    yield w
    w.stop()

N = 500
@pytest.fixture(scope="module")
def clean():
    def do_clean():
        for n in range(N):
            pathlib.Path(f"file{n}.canary").unlink(missing_ok=True)
    do_clean()
    yield
    do_clean()

@pytest.mark.parametrize("n", range(N))
def test(n, wM, clean):
    bar.send_with_options(kwargs={"s": str(n)}, delay=10)
    wM.broker.join(queue_name="default")
    wM.join()
    assert pathlib.Path(f"file{n}.canary").exists()
```
This can be run with pytest using `pytest -x`. Note that it can be quite rare on some environments, so a single run may not be sufficient. parallelizing with `-n` seems to help reproducing this, but it's not required to reporduce.

To better illustrate investigate the issue I've added some logs in the redis broker join operation:
```diff
diff --git a/dramatiq/brokers/redis.py b/dramatiq/brokers/redis.py   
index 0f37313..838595d 100644
--- a/dramatiq/brokers/redis.py
+++ b/dramatiq/brokers/redis.py
@@ -233,7 +233,10 @@ class RedisBroker(Broker):
 
             size = 0
             for name in (queue_name, dq_name(queue_name)):
-                size += self.do_qsize(name)
+                current = self.do_qsize(name)
+                self.logger.debug('queue %s size %s', name, current)
+                size += current
+            self.logger.debug('join total size %s', size)
 
             if size == 0:
                 return
```

with this change in place a failure logs the following:
```
===================================================================== FAILURES ======================================================================
_____________________________________________________________________ test[300] _____________________________________________________________________
[gw1] win32 -- Python 3.8.12 path\to\python.exe

n = 300, wM = <dramatiq.worker.Worker object at 0x0000012105930070>, clean = None

    @pytest.mark.parametrize("n", range(500))
    def test(n, wM, clean):
        bar.send_with_options(kwargs={"s": str(n)}, delay=10)
        wM.broker.join(queue_name="default")
        wM.join()
>       assert pathlib.Path(f"file{n}.canary").exists()
E       AssertionError: assert False
E        +  where False = <bound method Path.exists of WindowsPath('file300.canary')>()
E        +    where <bound method Path.exists of WindowsPath('file300.canary')> = WindowsPath('file300.canary').exists
E        +      where WindowsPath('file300.canary') = <class 'pathlib.Path'>('file300.canary')
E        +        where <class 'pathlib.Path'> = pathlib.Path

tests\a_test.py:70: AssertionError
----------------------------------------------------------------- Captured log call -----------------------------------------------------------------
DEBUG    dramatiq.broker.RedisBroker:redis.py:184 Enqueueing message '0619fc48-c787-4125-999e-8d439321e0ed' on queue 'default.DQ'.
DEBUG    dramatiq.broker.RedisBroker:redis.py:238 queue default size 0
DEBUG    dramatiq.broker.RedisBroker:redis.py:238 queue default.DQ size 1
DEBUG    dramatiq.broker.RedisBroker:redis.py:240 join total size 1
DEBUG    dramatiq.worker.ConsumerThread(default.DQ):worker.py:320 Pushing message '0619fc48-c787-4125-999e-8d439321e0ed' onto delay queue.
DEBUG    dramatiq.broker.RedisBroker:redis.py:184 Enqueueing message '0619fc48-c787-4125-999e-8d439321e0ed' on queue 'default'.
DEBUG    dramatiq.broker.RedisBroker:redis.py:238 queue default size 0
DEBUG    dramatiq.worker.ConsumerThread(default.DQ):worker.py:350 Acknowledging message '0619fc48-c787-4125-999e-8d439321e0ed'.
DEBUG    dramatiq.broker.RedisBroker:redis.py:238 queue default.DQ size 0
DEBUG    dramatiq.broker.RedisBroker:redis.py:240 join total size 0
============================================================== short test summary info ==============================================================
```

The race condition seems to be the following:
- the message delay has elapsed and processing of the message is in progress in the worker thread
- in the main thread the default queue size gets checked and returns 0
- the worker thread pushes the message to the default queue, then acks the message in the delay queue
- in the main thread the delay queue size gets checked and returns 0
- the join returns, but a message is present in the default queue.


The fix seems to be very easy, and it's just to swap in which order the queues are checked.
```diff
diff --git a/dramatiq/brokers/redis.py b/dramatiq/brokers/redis.py
index 0f37313..ce4aa98 100644
--- a/dramatiq/brokers/redis.py
+++ b/dramatiq/brokers/redis.py
@@ -232,7 +232,7 @@ class RedisBroker(Broker):
                 raise QueueJoinTimeout(queue_name)

             size = 0
-            for name in (queue_name, dq_name(queue_name)):
+            for name in (dq_name(queue_name), queue_name):
                 size += self.do_qsize(name)

             if size == 0:
```

This this change in place I've run the test above 20 times without any failure.

I'm opening a PR with the suggested fix

--- a/dramatiq/brokers/redis.py
+++ b/dramatiq/brokers/redis.py
@@ -231,11 +231,7 @@ class RedisBroker(Broker):
             if deadline and time.monotonic() >= deadline:
                 raise QueueJoinTimeout(queue_name)
 
-            size = 0
-            # The delay queue needs to be checked before the normal queue.
-            # See https://github.com/Bogdanp/dramatiq/issues/480 for details.
-            for name in (dq_name(queue_name), queue_name):
-                size += self.do_qsize(name)
+            size = self.do_qsize(queue_name)
 
             if size == 0:
                 return

--- a/dramatiq/brokers/redis.py
+++ b/dramatiq/brokers/redis.py
@@ -232,7 +232,9 @@ class RedisBroker(Broker):
                 raise QueueJoinTimeout(queue_name)
 
             size = 0
-            for name in (queue_name, dq_name(queue_name)):
+            # The delay queue needs to be checked before the normal queue.
+            # See https://github.com/Bogdanp/dramatiq/issues/480 for details.
+            for name in (dq_name(queue_name), queue_name):
                 size += self.do_qsize(name)
 
             if size == 0:

--- a/dramatiq/brokers/redis/dispatch.lua
+++ b/dramatiq/brokers/redis/dispatch.lua
@@ -229,8 +229,17 @@ elseif command == "purge" then
     redis.call("del", queue_full_name, queue_acks, queue_messages, xqueue_full_name, xqueue_messages)
 
 
--- Used in tests to determine the size of the queue.
+-- Used in tests to determine the size of the queue and its connected delay queue.
 elseif command == "qsize" then
-    return redis.call("hlen", queue_messages) + redis.call("scard", queue_acks)
+    local q_aks = acks .. "." .. queue_canonical_name
+    local q_fn = namespace .. ":" .. queue_canonical_name
+    local q_msgs = q_fn .. ".msgs"
+    local dq_q_aks = q_aks .. ".DQ"
+    local dq_q_msgs = q_fn .. ".DQ" .. ".msgs"
+
+    return redis.call("hlen", dq_q_msgs) +
+           redis.call("scard", dq_q_aks) +
+           redis.call("hlen", q_msgs) +
+           redis.call("scard", q_aks)
 
 end

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -496,40 +496,28 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
     """
     test for issue https://github.com/Bogdanp/dramatiq/issues/480
 
-    this test was triggered a race condition when ``qsize`` returned only the size of the
-    requested queue, not also of the delay queue.
+    this test verifies that do_qsize counts all messages in the queue and delay queue at the same
+    time, removing the race condition mentioned in the issue
     """
-    old_qsize = redis_broker.do_qsize
-    events = defaultdict(threading.Event)
-
-    class M(dramatiq.Middleware):
-        def before_enqueue(self, broker, message, delay):
-            if not delay:
-                events["default"].set()
-                events["default_done"].wait(2)
-
-        def after_ack(self, broker, message):
-            if message.options.get("eta"):
-                events[dq_name("default")].set()
-
-    def qsize(qn):
-        events[qn].wait(2)
-        events[qn].clear()
-        res = old_qsize(qn)
-        events[f"{qn}_done"].set()
-        return res
-
-    redis_broker.add_middleware(M())
-    redis_broker.do_qsize = qsize
     called = False
+    size = []
 
     @dramatiq.actor
     def go():
+        go_again.send_with_options(delay=50)
+        size.append(redis_broker.do_qsize('default'))  # go ack + go msg + go_again msg
+        size.append(redis_broker.do_qsize(dq_name('default')))  # does the same
+
+    @dramatiq.actor
+    def go_again():
         nonlocal called
         called = True
+        size.append(redis_broker.do_qsize('default'))  # go_again msg + go_again ack
+        size.append(redis_broker.do_qsize(dq_name('default')))  # does the same
 
-    go.send_with_options(delay=50)
+    go.send()
     redis_broker.join("default")
     redis_worker.join()
 
     assert called
+    assert size == [3, 3, 2, 2]

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -490,7 +490,7 @@ def test_redis_consumer_nack_does_not_raise_on_missing_id(redis_worker):
     consumer.nack(message)
 
 
-def test_redis_join_race_condition(redis_broker, redis_worker):
+def test_redis_join_race_condition(redis_broker):
     """
     test for issue https://github.com/Bogdanp/dramatiq/issues/480
 
@@ -503,7 +503,9 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
     @dramatiq.actor
     def go():
         go_again.send_with_options(delay=1000)
-        size.append(redis_broker.do_qsize("default"))  # go ack + go msg + go_again msg
+        time.sleep(0.25)
+        # go ack + go msg + go_again dq.msg + go_again db.ack
+        size.append(redis_broker.do_qsize("default"))
         size.append(redis_broker.do_qsize(dq_name("default")))  # does the same
 
     @dramatiq.actor
@@ -513,9 +515,10 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
         size.append(redis_broker.do_qsize("default"))  # go_again msg + go_again ack
         size.append(redis_broker.do_qsize(dq_name("default")))  # does the same
 
-    go.send()
-    redis_broker.join("default")
-    redis_worker.join()
+    with worker(redis_broker, worker_timeout=50, worker_threads=1) as redis_worker:
+        go.send()
+        redis_broker.join("default")
+        redis_worker.join()
 
     assert called
-    assert size == [3, 3, 2, 2]
+    assert size == [4, 4, 2, 2]

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -493,7 +493,12 @@ def test_redis_consumer_nack_does_not_raise_on_missing_id(redis_worker):
 
 
 def test_redis_join_race_condition(redis_broker, redis_worker):
-    "test for issue https://github.com/Bogdanp/dramatiq/issues/480"
+    """
+    test for issue https://github.com/Bogdanp/dramatiq/issues/480
+
+    this test was triggered a race condition when ``qsize`` returned only the size of the
+    requested queue, not also of the delay queue.
+    """
     old_qsize = redis_broker.do_qsize
     events = defaultdict(threading.Event)
 

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -1,4 +1,6 @@
+import threading
 import time
+from collections import defaultdict
 from unittest import mock
 
 import pytest
@@ -488,3 +490,41 @@ def test_redis_consumer_nack_does_not_raise_on_missing_id(redis_worker):
         options={"redis_message_id": "XXXXXXXXX"}
     )  # Bogus ID
     consumer.nack(message)
+
+
+def test_redis_join_race_condition(redis_broker, redis_worker):
+    "test for issue https://github.com/Bogdanp/dramatiq/issues/480"
+    old_qsize = redis_broker.do_qsize
+    events = defaultdict(threading.Event)
+
+    class M(dramatiq.Middleware):
+        def before_enqueue(self, broker, message, delay):
+            if not delay:
+                events['default'].set()
+                events['default_done'].wait(2)
+
+        def after_ack(self, broker, message):
+            if message.options.get("eta"):
+                events[dq_name('default')].set()
+
+    def qsize(qn):
+        events[qn].wait(2)
+        events[qn].clear()
+        res = old_qsize(qn)
+        events[f'{qn}_done'].set()
+        return res
+
+    redis_broker.add_middleware(M())
+    redis_broker.do_qsize = qsize
+    called = False
+
+    @dramatiq.actor
+    def go():
+        nonlocal called
+        called = True
+
+    go.send_with_options(delay=50)
+    redis_broker.join('default')
+    redis_worker.join()
+
+    assert called

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -500,18 +500,18 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
     class M(dramatiq.Middleware):
         def before_enqueue(self, broker, message, delay):
             if not delay:
-                events['default'].set()
-                events['default_done'].wait(2)
+                events["default"].set()
+                events["default_done"].wait(2)
 
         def after_ack(self, broker, message):
             if message.options.get("eta"):
-                events[dq_name('default')].set()
+                events[dq_name("default")].set()
 
     def qsize(qn):
         events[qn].wait(2)
         events[qn].clear()
         res = old_qsize(qn)
-        events[f'{qn}_done'].set()
+        events[f"{qn}_done"].set()
         return res
 
     redis_broker.add_middleware(M())
@@ -524,7 +524,7 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
         called = True
 
     go.send_with_options(delay=50)
-    redis_broker.join('default')
+    redis_broker.join("default")
     redis_worker.join()
 
     assert called

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -1,6 +1,4 @@
-import threading
 import time
-from collections import defaultdict
 from unittest import mock
 
 import pytest
@@ -505,15 +503,15 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
     @dramatiq.actor
     def go():
         go_again.send_with_options(delay=50)
-        size.append(redis_broker.do_qsize('default'))  # go ack + go msg + go_again msg
-        size.append(redis_broker.do_qsize(dq_name('default')))  # does the same
+        size.append(redis_broker.do_qsize("default"))  # go ack + go msg + go_again msg
+        size.append(redis_broker.do_qsize(dq_name("default")))  # does the same
 
     @dramatiq.actor
     def go_again():
         nonlocal called
         called = True
-        size.append(redis_broker.do_qsize('default'))  # go_again msg + go_again ack
-        size.append(redis_broker.do_qsize(dq_name('default')))  # does the same
+        size.append(redis_broker.do_qsize("default"))  # go_again msg + go_again ack
+        size.append(redis_broker.do_qsize(dq_name("default")))  # does the same
 
     go.send()
     redis_broker.join("default")

--- a/tests/test_redis.py
+++ b/tests/test_redis.py
@@ -502,7 +502,7 @@ def test_redis_join_race_condition(redis_broker, redis_worker):
 
     @dramatiq.actor
     def go():
-        go_again.send_with_options(delay=50)
+        go_again.send_with_options(delay=1000)
         size.append(redis_broker.do_qsize("default"))  # go ack + go msg + go_again msg
         size.append(redis_broker.do_qsize(dq_name("default")))  # does the same
 

